app/embedder.py

```diff
@@ -56,62 +54,82 @@ def __init__(self, bert: str):
-        l_matrix = matrix.tolist()
-        e = Embedding.make(dict(bert=self.bert, seq=seq, embedding=l_matrix))
-        e.m.save()
-        self.logger.info("Embedding matrix stored in MongoDB.")
-        return matrix
-
-    def _buildEmbedding(self, seq: str) -> np.array:
-        tokens = self.tokenizer.tokenize(seq)
-        if len(tokens) > MAX_SEQ_LENGTH - 2:
-            tokens = tokens[:(MAX_SEQ_LENGTH - 2)]
-
-        # CLS and SEP is a relic of multi-sentence pre-training
-        tokens = ["[CLS]"] + tokens + ["[SEP]"]
-
-        # segment_ids refers again to multi-sentence pre-training
-        # TODO: maybe scrap
-        segment_ids = [0] * len(tokens)
-
-        # per word, get IDs from vocab file
-        # e.g. [980234, 8792450, 132, 5002]
-        input_ids = self.tokenizer.convert_tokens_to_ids(tokens)
-
-        # The mask has 1 for real tokens and 0 for padding tokens. Only real
-        # tokens are attended to.
-        # e.g. [1, 1, 1, 1]
-        input_mask = [1] * len(input_ids)
-
-        # Zero-pad up to the sequence length.
-        padding = [0] * (MAX_SEQ_LENGTH - len(input_ids))  # [0, 0, 0, ...]
-        input_ids += padding  # [980234, 8792450, 132, 5002, 0, 0, 0, ...]
-        input_mask += padding  # [1, 1, 1, 1, 0, ...]
-        segment_ids += padding  # [0, 0, 0, ...]
-
-        assert len(input_ids) == MAX_SEQ_LENGTH
-        assert len(input_mask) == MAX_SEQ_LENGTH
-        assert len(segment_ids) == MAX_SEQ_LENGTH
-
-        bert_inputs = dict(input_ids=tf.expand_dims(input_ids, 0),
-                           input_mask=tf.expand_dims(input_mask, 0),
-                           segment_ids=tf.expand_dims(segment_ids, 0))
+        for i, seq in enumerate(seqs):
+            obj = Embedding.query.get(bert=self.bert, seq=seq)
+            if obj:
+                # convert list back from list(floats) to np.array
+                matrix = np.array(obj.embedding)
+                self.logger.debug(matrix)
+                result[seq] = matrix
+            else:
+                undone_seqs.add(seq)
+
+        num_done = sum(0 if r is None else 1 for r in result)
+        self.logger.info("Using %d of %d embedding matrices fetched from MongoDB." %
+                         (num_done, len(seqs)))
+        if len(undone_seqs) == 0:
+            return result
+
+        self.logger.info("Building %d embedding matrics with TensorFlow..." %
+                         (len(undone_seqs)))
+        done_seqs = self._buildEmbedding(undone_seqs)
+
+        for seq, matrix in done_seqs.items():
+            result[seq] = matrix
+            # convert npArray to list for storage in MongoDB
+            e = Embedding(bert=self.bert, seq=seq, embedding=matrix.tolist())
+        session.flush()
+        self.logger.info("Stored %d embedding matrices in MongoDB." % len(done_seqs))
+        return result
+
+    def _buildEmbedding(self, seqs: Set[str]) -> Dict[str, np.array]:
```

Comment:
Let's keep a sub-method for the singular embedding getter, and add a new one for doing the multi-sequence processing.

Reply:
I just removed the single-seq API per your other comment and this would lead to a lot of code duplication, I think.

Reply:
SGTM